Cannot change CSS scrollbar-width from none to auto

nmoucht opened the pull request:
#### 3bc0333b6ac70ba56b367cdde01ad6d15fb10a3f
<pre>
Cannot change CSS scrollbar-width from none to auto
<a href="https://bugs.webkit.org/show_bug.cgi?id=297501">https://bugs.webkit.org/show_bug.cgi?id=297501</a>
<a href="https://rdar.apple.com/158637000">rdar://158637000</a>

Reviewed by NOBODY (OOPS!).

Fix issue where going from scrollbar-width: none to auto on iOS wouldn&apos;t work.
Make use of RemoteScrollbarsController on iOS to properly send scrollbar width
updates to the UI process when changed via style.

* LayoutTests/fast/scrolling/ios/scrollbar-width-none-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-width-none.html: Added.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::horizontalNativeScrollbarVisibility const):
(WebCore::LocalFrameView::verticalNativeScrollbarVisibility const):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::operator&lt;&lt;):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::setScrollbarWidth):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameHorizontalScrollbarVisibility const):
(WebCore::ScrollingTree::mainFrameVerticalScrollbarVisibility const):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::nativeScrollbarVisibility):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::mainFrameHorizontalScrollbarVisibility const):
(WebKit::RemoteScrollingCoordinatorProxy::mainFrameVerticalScrollbarVisibility const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateAfterChildren):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::ensureScrollbarsController const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:
</pre><!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3bc0333b6ac70ba56b367cdde01ad6d15fb10a3f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/117964 "2 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/37642 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/28278 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/124111 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/69999 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/e4be756a-f8e2-4a7d-b3b8-753144d39b25) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/119842 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/38335 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/46224 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/89517 "Passed tests") | [💥 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/59126 "An unexpected error occured. Recent messages:Printed configuration") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/b609e6c9-21f1-4585-a45f-344cf720ee8b) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/120916 "Passed tests") | [❌ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/30531 "Found 2 new test failures: fast/scrolling/ios/scrollbar-hiding-iframes.html fast/scrolling/ios/scrollbar-hiding.html (failure)") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/105778 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/70010 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/328fae81-f4b7-4ce6-b84f-fe42cc45a40f) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/29598 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/23894 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/67774 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/99953 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/24072 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/127192 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/44867 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/33802 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/98183 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/45228 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/102002 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/97971 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/43376 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/21357 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/41301 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/44739 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/50413 "Built successfully") | | | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/44199 "Build is in progress. Recent messages:Running configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/47544 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/45888 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

Ahmad-S792 commented on LayoutTests/fast/scrolling/ios/scrollbar-width-none.html:
```diff
@@ -0,0 +1,86 @@
+            height: 200%;
+        }
+    </style>
+    <script src="../../../resources/js-test-pre.js"></script>
```
`js-test.js` instead of `-pre` and `-post`.

smfr commented on Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
```diff
@@ -1414,7 +1414,6 @@ RefPtr<WebCore::ScrollingCoordinator> WebChromeClient::createScrollingCoordinato
 
 #endif
 
-#if PLATFORM(MAC)
```
Could you explain in the commit message why we're now using `RemoteScrollbarsController` on non-macOS platforms too? It could use some more detail.

nmoucht replied:
Fixed.

nmoucht committed:
Cannot change CSS scrollbar-width from none to auto
https://bugs.webkit.org/show_bug.cgi?id=297501
rdar://158637000

Reviewed by NOBODY (OOPS!).

Fix issue where going from scrollbar-width: none to auto on iOS wouldn't work.
Make use of RemoteScrollbarsController on iOS to properly send scrollbar width
updates to the UI process when changed via style.

* LayoutTests/fast/scrolling/ios/scrollbar-width-none-expected.txt: Added.
* LayoutTests/fast/scrolling/ios/scrollbar-width-none.html: Added.
* Source/WebCore/page/LocalFrameView.cpp:
(WebCore::LocalFrameView::horizontalNativeScrollbarVisibility const):
(WebCore::LocalFrameView::verticalNativeScrollbarVisibility const):
* Source/WebCore/page/scrolling/ScrollingCoordinatorTypes.cpp:
(WebCore::operator<<):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::setScrollbarWidth):
* Source/WebCore/page/scrolling/ScrollingTree.cpp:
(WebCore::ScrollingTree::mainFrameHorizontalScrollbarVisibility const):
(WebCore::ScrollingTree::mainFrameVerticalScrollbarVisibility const):
* Source/WebCore/page/scrolling/ScrollingTree.h:
* Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
* Source/WebCore/platform/Scrollbar.cpp:
(WebCore::Scrollbar::nativeScrollbarVisibility):
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _updateScrollViewForTransaction:]):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.cpp:
(WebKit::RemoteScrollingCoordinatorProxy::mainFrameHorizontalScrollbarVisibility const):
(WebKit::RemoteScrollingCoordinatorProxy::mainFrameVerticalScrollbarVisibility const):
* Source/WebKit/UIProcess/RemoteLayerTree/RemoteScrollingCoordinatorProxy.h:
* Source/WebKit/UIProcess/RemoteLayerTree/ios/ScrollingTreeScrollingNodeDelegateIOS.mm:
(WebKit::ScrollingTreeScrollingNodeDelegateIOS::commitStateAfterChildren):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.cpp:
(WebKit::WebChromeClient::ensureScrollbarsController const):
* Source/WebKit/WebProcess/WebCoreSupport/WebChromeClient.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.h:
* Source/WebKit/WebProcess/WebPage/RemoteLayerTree/RemoteScrollbarsController.mm:

karlcow commented on Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:
```diff
@@ -132,6 +132,8 @@ class WEBCORE_EXPORT ScrollingTreeScrollingNode : public ScrollingTreeNode {
     OverscrollBehavior horizontalOverscrollBehavior() const { return m_scrollableAreaParameters.horizontalOverscrollBehavior; }
     OverscrollBehavior verticalOverscrollBehavior() const { return m_scrollableAreaParameters.verticalOverscrollBehavior; }
 
+    NativeScrollbarVisibility horizontalScrollbarVisibility() const { return m_scrollableAreaParameters.horizontalNativeScrollbarVisibility; }
+    NativeScrollbarVisibility verticalScrollbarVisibility() const { return m_scrollableAreaParameters.verticalNativeScrollbarVisibility; }
```
Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:136: error: [build/webcore_export] Inline functions should not be in classes annotated with WEBCORE_EXPORT. Remove the macro from the class and apply it to each appropriate method, or move the inline function definition out-of-line.
Source/WebCore/page/scrolling/ScrollingTreeScrollingNode.h:137: error: [build/webcore_export] Inline functions should not be in classes annotated with WEBCORE_EXPORT. Remove the macro from the class and apply it to each appropriate method, or move the inline function definition out-of-line.

karlcow reviewed (commented):
These two tests are failing 

```
fast/scrolling/ios/scrollbar-hiding-iframes.html
fast/scrolling/ios/scrollbar-hiding.html 
```
